src/visualization/makie_visualization.jl

```diff
@@ -0,0 +1,373 @@
+
+  @unpack variable_names = pd
+
+  semi = sol.prob.p
+  @unpack mesh = semi
+  @assert ndims(mesh) == 2
+
+  fig = Makie.Figure()
+
+  # set up menu and toggle switches
+  options = [zip(variable_names, 1:length(variable_names))...]
+  menu = Makie.Menu(fig, options=options)
+  toggle_solution_mesh = Makie.Toggle(fig, active=true)
+  toggle_mesh = Makie.Toggle(fig, active=true)
+  fig[1, 1] = Makie.vgrid!(
+      Makie.Label(fig, "Solution field", width=nothing), menu,
+      Makie.Label(fig, "Solution mesh visible"), toggle_solution_mesh,
+      Makie.Label(fig, "Mesh visible"), toggle_mesh;
+      tellheight=false, width = 200
+  )
+
+  ax = Makie.LScene(fig[1, 2], scenekw=(show_axis = false,))
+
+  # interactive menu variable_to_plot
+  menu.selection[] = variable_to_plot_in
+  menu.i_selected[] = variable_to_plot_in
+
+  # these lines get re-run whenever variable_to_plot[] is updated
+  plotting_mesh = Makie.@lift(generate_plotting_triangulation(getindex(pd, variable_names[$(menu.selection)])))
+  solution_z = Makie.@lift(getindex.($plotting_mesh.position, 3))
+
+  Makie.mesh!(ax, plotting_mesh, color=solution_z, nvisnodes=nvisnodes)
+
+  # mesh overlay: we plot a mesh both on top of and below the solution contours
+  wire_points = Makie.@lift(generate_plotting_wireframe(getindex(pd, variable_names[$(menu.selection)])))
+  wire_mesh_top = Makie.lines!(ax, wire_points, color=:white)
+  wire_mesh_bottom = Makie.lines!(ax, wire_points, color=:white)
+  Makie.translate!(wire_mesh_top, 0, 0, 1e-3)
+  Makie.translate!(wire_mesh_bottom, 0, 0, -1e-3)
+
+  # draw a flattened mesh wireframe below the solution
+  function compute_z_offset(solution_z)
+      zmin = minimum(solution_z)
+      zrange = (x->x[2]-x[1])(extrema(solution_z))
+      return zmin - .25*zrange
+  end
+  z_offset = Makie.@lift(compute_z_offset($solution_z))
+  get_flat_points(wire_points, z_offset) = [Makie.Point(point.data[1:2]..., z_offset) for point in wire_points]
+  flat_wire_points = Makie.@lift get_flat_points($wire_points, $z_offset)
+  wire_mesh_flat = Makie.lines!(ax, flat_wire_points, color=:black)
+
+  # reset colorbar each time solution changes
+  Makie.Colorbar(fig[1, 3], limits = Makie.@lift(extrema($solution_z)), colormap = :viridis, flipaxis = false)
+
+  # syncs the toggle to the mesh
+  Makie.connect!(wire_mesh_top.visible, toggle_solution_mesh.active)
+  Makie.connect!(wire_mesh_bottom.visible, toggle_solution_mesh.active)
+  Makie.connect!(wire_mesh_flat.visible, toggle_mesh.active)
+
+  # On OSX, shift-command-4 for screenshots triggers a constant "up-zoom".
+  # To avoid this, we remap up-zoom to the right shift button instead.
+  Makie.cameracontrols(ax.scene).attributes[:up_key][] = Makie.Keyboard.right_shift
+
+  # typing this pulls up the figure (similar to display(plot!()) in Plots.jl)
+  fig
+end
+
+# ================== new Makie plot recipes ====================
+
+@Makie.recipe(TrixiHeatmap, plot_data_series) do scene
+  Makie.Theme(;)
+end
```

Comment:
I have zero understanding what this method does - obviously some black behind-the-scenes magic. Could you add a short comment telling what this does or why it's necessary?

Reply:
I've added some comments, but much of it is just Makie black magic. Hope it's clearer now!

Reply:
Thanks!